### ISSUE TITLE
Change order in USAGE when %*SUB-MAIN-OPTS<named-anywhere> is set

### ISSUE THREAD

**Issue by Juerd** [rakudo/rakudo#2246]
## The Problem

Sub-commands in USAGE output are ugly and not how they are intended to be used.

## Expected Behavior

```
  zef build [--force|--force-build] [<paths> ...] -- Run Build.pm
  zef install [--depends] [--test-depends] [--build-depends] [--test] [--fetch] [--build] [--force] [--force-resolve] [--force-fetch] [--force-extract] [--force-build] [--force-test] [--force-install] [--dry] [--update] [--upgrade] [--deps-only] [--serial] [--exclude=<Any>] [--to|--install-to=<Any>] [<wants> ...] -- Install
  zef uninstall [--from|--uninstall-from=<Any>] [<identities> ...] -- Uninstall
```

## Actual Behavior

```
  zef [--force|--force-build] build [<paths> ...] -- Run Build.pm
  zef [--depends] [--test-depends] [--build-depends] [--test] [--fetch] [--build] [--force] [--force-resolve] [--force-fetch] [--force-extract] [--force-build] [--force-test] [--force-install] [--dry] [--update] [--upgrade] [--deps-only] [--serial] [--exclude=<Any>] [--to|--install-to=<Any>] install [<wants> ...] -- Install
  zef [--from|--uninstall-from=<Any>] uninstall [<identities> ...] -- Uninstall
```

## Rationale

The basic rationale is already described in PR #688, and in commit af145848 which fixes the parsing but not yet how USAGE generates its output by default.

It is likely that literal strings in signatures, especially when they occur as the first positional argument, are intended as subcommands. However, the default parsing of command line arguments in Perl 6 is awkward and does not allow subcommands to be specified before named arguments. That is fixed (well, sort of, because the default still fails to make sense) since the addition of %*SUB-MAIN-OPTS with the `named-anywhere` option (which implies that positionals can also go anywhere), but then there's still the issue of the automatically generated usage information.

My suggestion is to simply assume that anyone who sets `named-anywhere` in %*SUB-MAIN-OPTS will also love to see literal positional arguments before the named arguments in the usage string.

**Comment by Juerd** [comment]
A quick scan of the entire ecosystem shows that literal string positional arguments to MAIN are always the first argument defined, and I'm quite certain that without exception they are all intended as subcommands,

**Comment by ugexe** [comment]
fwiw zef doesn’t used %*SUB-MAIN-OPTS<named-anywhere> ( which came after the same functionality in zef  ) although it probably should

**Comment by Leont** [comment]
I agree this is sensible

**Comment by patzim** [comment]
Do I understand correctly, that, with the exception of the USAGE message, this already works as proposed? So the design decision that was blocking PR #688 is not blocking anymore?

**Comment by Juerd** [comment]
Yes, commit af14584 implements parsing as necessary for this feature, at least with %*SUB-MAIN-OPTS<named-anywhere>. This issue is specifically about the generated USAGE message, which is not in line with the most common (and I believe only) intended use case.

**Comment by lizmat** [comment]
Not sure when this was done, but I think we already have the intended behaviour.  Put this in a script "foo.p6":

    my %*SUB-MAIN-OPTS = :named-anywhere;
    multi sub MAIN("foo", :$bar) { }
    multi sub MAIN("bat", :$zip) { }

then run it with `perl6 foo.p6": you will get:

    Usage:
      foo.p6 foo [--bar=<Any>]
      foo.p6 bat [--zip=<Any>]

Then comment out the first line, and you'll get:

    Usage:
      foo.p6 [--bar=<Any>] foo
      foo.p6 [--zip=<Any>] bat

Please close the ticket if you concur.  If not, please let me know what the problem is  :-)

**Comment by Juerd** [comment]
It seems you implemented this in commit 0d1be77e. Thanks!
